gammapy/cube/psf_kernel.py

```diff
@@ -0,0 +1,266 @@
+    geom : `~gammapy.maps.MapGeom`
+        the target geometry. The PSF kernel will be centered on the spatial center.
+    factor : int
+        the oversample factor to compute the PSF
+    """
+
+    # prepare map and compute distances to map center
+    kernel_map, rads = _compute_kernel_separations(geom, factor)
+
+    vals = table_psf.evaluate(rad=rads)
+    norm = np.sum(vals).value
+
+    # loop over images and fill map
+    for img, idx in kernel_map.iter_by_image():
+        img += vals.reshape(img.shape).value/norm
+
+    # downsample the psf kernel map. Take the average
+    kernel_map = kernel_map.downsample(factor, preserve_counts=True)
+
+    return kernel_map
+
+
+def energy_dependent_table_psf_to_kernel_map(table_psf, geom, factor=4):
+    """Compute an energy dependent PSF kernel on a given MapGeom.
+
+    The PSF is estimated by oversampling defined by a given factor.
+
+    Parameters
+    ----------
+    table_psf : `~gammapy.irf.EnergyDependentTablePSF`
+        the input table PSF
+    geom : `~gammapy.maps.MapGeom`
+        the target geometry. The PSF kernel will be centered on the spatial centre.
+        the geometry axes should contain an energy MapAxis. The kernel will be
+        duplicated along other axes.
+    factor : int
+        the oversample factor to compute the PSF
+    """
+    # Find energy axis in geom
+    energy_axis = geom.get_axis_by_type('energy')
+    energy_idx = geom.axes.index(energy_axis)
+    energy_unit = u.Unit(energy_axis.unit)
+
+    # prepare map and compute distances to map center
+    kernel_map, rads = _compute_kernel_separations(geom, factor)
+
+    # loop over images
+    for img, idx in kernel_map.iter_by_image():
+        energy = energy_axis.center[idx[energy_idx]] * energy_unit
+        vals = table_psf.evaluate(energy=energy, rad=rads).reshape(img.shape)
+        img += vals.value / np.sum(vals).value
+
+    # downsample the psf kernel map. Take the average
+    kernel_map = kernel_map.downsample(factor, preserve_counts=True)
+
+    return kernel_map
+
+
+class PSFKernel(object):
+    """PSF kernel for `~gammapy.maps.Map`.
+
+    This is a container class to store a PSF kernel
+    that can be used to convolve `~gammapy.maps.WcsNDMap` objects.
+    It is usually computed from an EnergyDependentTablePSF
+
+    Parameters
+    ----------
+    psf_kernel_map : `~gammapy.maps.Map`
+        PSF kernel stored in a Map
+
+    Examples
+    --------
+
+    .. code:: python
+
+        import numpy as np
+        from gammapy.maps import WcsGeom, MapAxis
+        from astropy import units as u
+
+        # Define energy axis
+        energy_axis = MapAxis.from_edges(np.logspace(-1., 1., 4), unit='TeV', name='energy')
+
+        # Create WcsGeom
+        geom = WcsGeom.create(binsz=0.02*u.deg, width=10.0*u.deg, axes=[energy_axis])
+
+        # Extract EnergyDependentTablePSF from PSF IRF (here a PSF3D)
+        table_psf = psf.to_energy_dependent_table_psf(theta=0.5*u.deg)
```

Comment:
Can you fix this and make a minimal working example?
```
In [1]: %paste
        import numpy as np
        from gammapy.maps import WcsGeom, MapAxis
        from astropy import units as u

        # Define energy axis
        energy_axis = MapAxis.from_edges(np.logspace(-1., 1., 4), unit='TeV', name='energy')

        # Create WcsGeom
        geom = WcsGeom.create(binsz=0.02*u.deg, width=10.0*u.deg, axes=[energy_axis])

        # Extract EnergyDependentTablePSF from PSF IRF (here a PSF3D)
        table_psf = psf.to_energy_dependent_table_psf(theta=0.5*u.deg)

        psf_kernel = PSFKernel.from_table_psf(table_psf,geom, max_radius=1*u.deg)

        some_map_convolved = psf_kernel.apply(some_map)
## -- End pasted text --
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-1-0b05c82cb4ce> in <module>()
     10 
     11 # Extract EnergyDependentTablePSF from PSF IRF (here a PSF3D)
---> 12 table_psf = psf.to_energy_dependent_table_psf(theta=0.5*u.deg)
     13 
     14 psf_kernel = PSFKernel.from_table_psf(table_psf,geom, max_radius=1*u.deg)

NameError: name 'psf' is not defined
```

Reply:
Done. It works now.